### PR TITLE
Failed batch breakdown

### PR DESCRIPTION
**Pull request by Nosorogh**

### CODE REVIEW

File: src/Eventso.Subscription/Observing/Batch/BatchEventObserver.cs [hunk 1]
```diff
@@ -34,8 +36,12 @@ public sealed class BatchEventObserver<TEvent> : IObserver<TEvent>, IDisposable
             _messageHandlersRegistry = messageHandlersRegistry;
             _skipUnknown = skipUnknown;
 
-            _cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
-                consumer.CancellationToken);
+            _failedBufferProcessor = FailedBufferProcessorResolver.Resolve(
```
Comment (Nosorogh) [inline src/Eventso.Subscription/Observing/Batch/BatchEventObserver.cs:39]: looks like a thing that has to be injected, but `Buffer<TEvent>.BufferedEvent` is internal while the `BatchEventObserver` constructor is public

File: src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            _handler = handler;
+            _consumer = consumer;
+        }
+
+        public async Task Process(
+            IReadOnlyList<Buffer<TEvent>.BufferedEvent> bufferedEvents,
+            CancellationToken cancellationToken)
+        {
+            SingleEventCollection convertibleCollection = null;
```
Comment (deniskuzmin) [inline src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs:25]: if we got here then handling occured. that means that at least 1 not skipped event was in batch. so u will create SingleEventCollection anyway. so u can create empty collection, remove `is null` check on every iteration and use `SetEvent`

File: src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                var @event = bufferedEvent.Event;
+
+                if (convertibleCollection is null)
+                    convertibleCollection = new SingleEventCollection(@event);
+                else
+                    convertibleCollection.SetEvent(@event);
+
+                await _handler.Handle(convertibleCollection, cancellationToken);
+
+                _consumer.Acknowledge(@event);
```
Comment (deniskuzmin) [inline src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs:43]: nothing will acknowledge skipped events if retry handled successfully and skipped events were in the end of initial collection

File: src/Eventso.Subscription/Observing/Batch/BatchEventObserver.cs [hunk 4]
```diff
@@ -137,28 +157,31 @@ private async Task HandleBatch(PooledList<Buffer<TEvent>.BufferedEvent> events,
             }
             catch
             {
-                _consumer.Cancel();
+                await _failedBufferProcessor.Process(events, _cancellationTokenSource.Token);
```
[the remark this answers is not in the document]
Comment (Nosorogh) [reply src/Eventso.Subscription/Observing/Batch/BatchEventObserver.cs:160]: > `_consumer.Cancel()` logic is lost now

why? there is `throw;` after the breakdown call

File: src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs (hunk 3 above)
Comment (Nosorogh) [reply src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs:43]: fair enough. seems like I should ack every single event, including skipped ones

File: src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs (hunk 2 above)
Comment (Nosorogh) [reply src/Eventso.Subscription/Observing/Batch/ErrorHandling/BreakdownFailedBufferProcessor.cs:26]: I don't like it too, but I had to choose: complicated SingleEventCollection that supports zero size or complicated loop body. I tried both and decided that the second option is better)

**Review by deniskuzmin — APPROVED** [review]

**Pull request merged**